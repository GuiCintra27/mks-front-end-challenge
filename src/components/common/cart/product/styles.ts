import styled from "styled-components";

export const Container = styled.div`
  position: relative;

  padding: 1.9rem 2.3rem;
  width: 100%;
  height: 9.5rem;

  display: flex;
  align-items: center;
  gap: 2rem;

  border-radius: 0.8rem;
  background: ${({ theme }) => theme.colors.primary};
  box-shadow: -2px 2px 10px 0px rgba(0, 0, 0, 0.05);
  color: ${({ theme }) => theme.colors.onPrimary};

  .close {
    position: absolute;
    top: 0;
    right: 0;

    transform: translate(50%, -50%);
    cursor: pointer;
  }

  .description {
    display: flex;
    gap: 2rem;
    align-items: center;

    img {
      width: 7rem;
      max-height: 6rem;

      border-radius: 0.8rem;
    }

    h3 {
      width: 8.5rem;
      word-wrap: break-word;

      font-size: 1.3rem;
      line-height: 1.7rem; /* 130.769% */
    }
  }

  .price {
    color: ${({ theme }) => theme.colors.onPrimary};
    font-size: 1.4rem;
    font-weight: 700;
    line-height: 1.7rem; /* 121.429% */
  }
`;

export const QuantityWrapper = styled.div`
  width: 6rem;

  display: flex;
  flex-direction: column;
  gap: 0.4rem;

  font-size: 0.8rem;

  .button {
    padding: 0.5rem;
    height: 2.4rem;
    width: 100%;

    display: flex;
    justify-content: space-around;
    align-items: center;

    border: 1px solid #bfbfbf;
    border-radius: 0.4rem;

    span {
      font-size: 1.2rem;
      cursor: pointer;
    }
  }

  .separator {
    width: 0.2px;
    height: 1.15rem;

    background-color: #bfbfbf;
  }
`;
