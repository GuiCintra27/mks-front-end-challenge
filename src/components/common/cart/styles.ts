"use client";

import styled from "styled-components";

export const Container = styled.div`
  position: fixed;
  top: 0;
  right: 0;
  z-index: 1;

  height: 100vh;
  width: 48.6rem;

  background-color: ${({ theme }) => theme.colors.secondary};
  box-shadow: -5px 0px 6px 0px rgba(0, 0, 0, 0.13);

  .title-wrapper {
    margin: 3.6rem 2.2rem 7rem 4.7rem;

    display: flex;
    align-items: center;
    justify-content: space-between;

    h3 {
      max-width: 18rem;

      color: #fff;
      font-size: 2.7rem;
      font-weight: 700;
    }

    img {
      cursor: pointer;
      filter: invert(${({ theme }) => theme.invert});
    }
  }

  .products-list {
    margin: 0 5.4rem 3rem 4.7rem;
    padding-top: 1rem;
    padding-right: 1rem;
    height: 60%;
    overflow-y: scroll;

    display: flex;
    flex-direction: column;
    gap: 2.2rem;

    &::-webkit-scrollbar {
      display: none;
    }
  }

  .total {
    margin: 0 5.4rem 4rem 4.7rem;
    width: calc(100% - 10rem);

    display: flex;
    align-items: center;
    justify-content: space-between;

    color: #fff;
    font-size: 2.8rem;
    font-weight: 700;
    line-height: 1.5rem; /* 53.571% */
  }
`;
