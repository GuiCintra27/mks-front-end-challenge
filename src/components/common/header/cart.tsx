import Image from "next/image";
import styled from "styled-components";
import { useDispatch } from "react-redux";

import { toggleCart } from "@/components/infra/storage/cart-slice";

export function Cart() {
  const dispatch = useDispatch();
  return (
    <Container onClick={() => dispatch(toggleCart())}>
      <Image src="/icons/cart-icon.svg" width={19} height={18} alt="Cart" />
      <span>0</span>
    </Container>
  );
}

const Container = styled.div`
  padding: 1.3rem 1.5rem;

  display: flex;
  align-items: center;
  gap: 1.6rem;

  border-radius: 0.8rem;
  background: #fff;
  cursor: pointer;

  span {
    color: #000;
    font-size: 1.8rem;
    font-weight: 700;
    line-height: normal;
  }
`;
