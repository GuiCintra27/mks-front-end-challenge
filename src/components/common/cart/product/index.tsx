import { ProductModel } from "@/models/product";
import { Container, QuantityWrapper } from "./styles";

export function CartProduct({
  name,
  photo,
  price,
  quantity,
}: Pick<ProductModel, "name" | "photo" | "price"> & { quantity: number }) {
  return (
    <Container>
      <div className="description">
        <img src={photo} />
        <h3>{name}</h3>
      </div>
      <QuantityWrapper>
        <span>Qtd:</span>
        <div className="button">
          <span>-</span>
          <div className="separator" />
          <div>{quantity}</div>
          <div className="separator" />
          <span>+</span>
        </div>
      </QuantityWrapper>
      <p className="price">R$ {price}</p>
    </Container>
  );
}
