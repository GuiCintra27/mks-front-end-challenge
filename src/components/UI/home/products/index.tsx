
import { Product } from "./product";
import { Container } from "./styles";
import { ProductProps } from "@/models/product";

export function Products({products}: {products: ProductProps[]}) {
  return (
    <Container>
      {products.map((product) => (
        <Product {...product} key={product.id} />
      ))}
    </Container>
  );
}


