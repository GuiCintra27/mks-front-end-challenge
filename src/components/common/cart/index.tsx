"use client";

import Image from "next/image";
import { useDispatch } from "react-redux";

import { Container } from "./styles";
import { useAppSelector } from "@/hooks/useAppSelector";
import { toggleCart } from "@/components/infra/storage/cart-slice";
import { CartProduct } from "./product";

export function Cart() {
  const { isActive, products } = useAppSelector((state) => state.cart);
  const dispatch = useDispatch();

  if (!isActive) return null;


  return (
    <Container>
      <div className="title-wrapper">
        <h3>Carrinho de compras</h3>
        <Image
          src="/icons/close-cart-icon.svg"
          width={38}
          height={38}
          alt="Fechar"
          onClick={() => dispatch(toggleCart())}
        />
      </div>
      <div className="products-list">
        {products.map((product) => (
          <CartProduct
            id={product.id}
            name={product.name}
            price={product.price * product.quantity}
            quantity={product.quantity}
            photo={product.photo}
            key={product.id}
          />
        ))}
      </div>
      <div className="total">
        <span>Total</span>
        <span>R$ {products.reduce((acc, product) => acc + product.price, 0)}</span>
      </div>
    </Container>
  );
}
