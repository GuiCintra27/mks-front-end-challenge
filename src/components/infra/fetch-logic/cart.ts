import { AnyAction, Dispatch } from "redux";

import { ProductModel } from "@/models/product";
import { confirmAlert, successToast } from "@/components/UI/alerts";
import { clearCart, insertItem, removeItem } from "../storage/cart-slice";

export function addOnCart({
  dispatch,
  productData,
}: {
  dispatch: Dispatch<AnyAction>;
  productData:
    | ProductModel
    | Pick<ProductModel, "id" | "name" | "price" | "photo">;
}) {
  dispatch(
    insertItem({
      ...productData,
    })
  );
}

export function removeFromCart({
  dispatch,
  id,
  quantity,
  removeAll,
}: {
  dispatch: Dispatch<AnyAction>;
  id: number;
  quantity: number;
  removeAll?: boolean;
}) {
  if (quantity === 1 || removeAll) {
    confirmAlert(
      "Você tem certeza?",
      "Este produto será removido do carrinho",
      {
        confirm: "Continuar",
        cancel: "Cancelar",
      }
    ).then((response) => {
      if (response.isConfirmed) {
        dispatch(removeItem({ id, removeAll }));
      }
    });
  } else dispatch(removeItem({ id }));
}

export function finalizePurchase({
  dispatch,
  products,
}: {
  dispatch: Dispatch<AnyAction>;
  products: Pick<ProductModel, "id" | "name" | "price" | "photo">[];
}) {
  if (products.length === 0) return;

  confirmAlert("Tem certeza?", "A compra será finalizada", {
    confirm: "Continuar",
    cancel: "Cancelar",
  }).then((response) => {
    if (response.isConfirmed) {
      dispatch(clearCart());
      successToast("Sua compra foi finalizada com sucesso");
    }
  });
}
