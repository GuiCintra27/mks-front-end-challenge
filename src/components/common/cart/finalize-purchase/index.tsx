import { useDispatch } from "react-redux";

import { Container } from "./styles";
import { ProductModel } from "@/models/product";
import { finalizePurchase } from "@/components/infra/fetch-logic/cart";

export function FinalizePurchase({products}: {products: Pick<ProductModel, "id" | "name" | "price" | "photo">[]}) {
  const dispatch = useDispatch();

  return <Container onClick={() => finalizePurchase({dispatch, products})}>Finalizar compra</Container>;
}
