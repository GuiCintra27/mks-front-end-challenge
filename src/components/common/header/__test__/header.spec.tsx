import { fireEvent, render, screen } from "@testing-library/react";

import { Cart } from "../cart";
import { ThemeProvider } from "@/theme/theme";
import { CartProduct } from "../../cart/product";
import { ReduxProvider } from "@/lib/reduxProvider";
import { Product } from "@/components/UI/home/products/product";
import { QueryClientProvider } from "@/lib/queryClientProvider";
import { ChangeThemeButton } from "../changeThemeButton";

const productData = {
  id: 1,
  name: "name",
  photo: "photo",
  price: 10,
  description: "description",
  createdAt: new Date().toDateString(),
  updatedAt: new Date().toDateString(),
  brand: "brand",
};

describe("Header", () => {
  describe("Cart", () => {
    function renderComponent({
      children,
      quantity = 1,
    }: { children?: React.ReactNode; quantity?: number } = {}) {
      return render(
        <QueryClientProvider>
          <ReduxProvider>
            <ThemeProvider>
              <Cart />
              {new Array(quantity).fill(0).map((_, index) => (
                <Product key={index} {...productData} id={index} />
              ))}
              {children}
            </ThemeProvider>
          </ReduxProvider>
        </QueryClientProvider>
      );
    }

    it("Initial value should be 0", () => {
      const { getByTestId } = render(
        <ReduxProvider>
          <Cart />
        </ReduxProvider>
      );

      const span = getByTestId("cart-items-count");

      expect(span.textContent).toBe("0");
    });
    describe("When a new item is added", () => {
      it("The value should increase by 1", () => {
        const { getByTestId } = renderComponent();

        const buyButton = getByTestId("buy-button-0");

        fireEvent.click(buyButton);

        const span = getByTestId("cart-items-count");

        expect(span.textContent).toBe("1");
      });
      it("If the same item already in the cart, the cart value must remain the same", () => {
        const { getByTestId } = renderComponent();

        const buyButton = getByTestId("buy-button-0");

        fireEvent.click(buyButton);
        fireEvent.click(buyButton);

        const span = getByTestId("cart-items-count");

        expect(span.textContent).toBe("1");
      });
      describe("When two different itens are added", () => {
        it("The cart value must increase by 2", () => {
          const { getByTestId } = renderComponent({ quantity: 2 });

          const buyButton = getByTestId("buy-button-0");
          const buyButton2 = getByTestId("buy-button-1");

          fireEvent.click(buyButton);
          fireEvent.click(buyButton2);

          const span = getByTestId("cart-items-count");

          expect(span.textContent).toBe("2");
        });
      });
    });

    describe("When an item is removed", () => {
      it("If was the last item, an alert should appear", () => {
        const { getByTestId } = renderComponent({
          children: <CartProduct {...productData} quantity={1} />,
        });

        const buyButton = getByTestId("buy-button-0");
        const decreaseButton = getByTestId("decrease-button-1");

        fireEvent.click(buyButton);

        fireEvent.click(decreaseButton);

        expect(screen.getByText("VocÃª tem certeza?")).toMatchSnapshot();
      });
    });
  });

  describe("Theme", () => {
    function renderComponent() {
      return render(
        <ReduxProvider>
          <ThemeProvider>
            <ChangeThemeButton />
            <Product {...productData} />
          </ThemeProvider>
        </ReduxProvider>
      );
    }

    it("Should change the theme when change theme button is clicked", async () => {
      const { getByTestId } = renderComponent();

      const changeThemeButton = getByTestId("change-theme-button");
      const productContainer = getByTestId("product-container");
      const style = window.getComputedStyle(productContainer);

      expect(style.backgroundColor).toBe("rgb(255, 255, 255)");

      fireEvent.click(changeThemeButton);

      const newProductContainer = getByTestId("product-container");
      const newStyle = window.getComputedStyle(newProductContainer);

      expect(newStyle.backgroundColor).toBe("rgb(24, 26, 42)");
    });
  });
});
