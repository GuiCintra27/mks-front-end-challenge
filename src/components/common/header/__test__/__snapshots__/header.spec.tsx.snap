// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Cart When an item is removed If was the last item, an alert should appear 1`] = `
<h2
  class="swal2-title"
  id="swal2-title"
  style="display: block;"
>
  VocÃª tem certeza?
</h2>
`;
